ksqldb-engine/src/main/java/io/confluent/ksql/format/DefaultFormatInjector.java

```diff
@@ -0,0 +1,190 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OF ANY KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+
+package io.confluent.ksql.format;
+
+import io.confluent.ksql.parser.KsqlParser.PreparedStatement;
+import io.confluent.ksql.parser.SqlFormatter;
+import io.confluent.ksql.parser.exception.ParseFailedException;
+import io.confluent.ksql.parser.properties.with.CreateSourceProperties;
+import io.confluent.ksql.parser.tree.CreateAsSelect;
+import io.confluent.ksql.parser.tree.CreateSource;
+import io.confluent.ksql.parser.tree.Statement;
+import io.confluent.ksql.properties.with.CommonCreateConfigs;
+import io.confluent.ksql.serde.FormatInfo;
+import io.confluent.ksql.statement.ConfiguredStatement;
+import io.confluent.ksql.statement.Injector;
+import io.confluent.ksql.util.ErrorMessageUtil;
+import io.confluent.ksql.util.KsqlConfig;
+import io.confluent.ksql.util.KsqlException;
+import io.confluent.ksql.util.KsqlStatementException;
+import java.util.Optional;
+
+/**
+ * An injector which injects the key and value formats into the supplied {@code statement}.
+ *
+ * <p>The key format is only injected if:
+ * <ul>
+ * <li>The statement is a CT/CS.</li>
+ * <li>The statement does not specify the FORMAT property in its WITH clause.</li>
+ * <li>The statement does not specify the KEY_FORMAT property in its WITH clause.</li>
+ * </ul>
+ *
+ * <p>Similarly, the value format is only injected if the above conditions are met,
+ * where the KEY_FORMAT property is replaced with the VALUE_FORMAT property accordingly.
+ *
+ * <p>If any of the above are not true then the {@code statement} is returned unchanged.
+ */
+public class DefaultFormatInjector implements Injector {
+
+  public DefaultFormatInjector() {
+  }
+
+  @SuppressWarnings("unchecked")
+  @Override
+  public <T extends Statement> ConfiguredStatement<T> inject(
+      final ConfiguredStatement<T> statement
+  ) {
+    if (featureFlagNotEnabled(statement)) {
+      validateLegacyFormatProperties(statement);
+    }
+
+    if (statement.getStatement() instanceof CreateSource) {
+      return handleCreateSource((ConfiguredStatement<CreateSource>) statement);
+    }
+
+    return statement;
+  }
+
+  @SuppressWarnings("unchecked")
+  private <T extends Statement> ConfiguredStatement<T> handleCreateSource(
+      final ConfiguredStatement<CreateSource> statement
+  ) {
+    try {
+      // Safe to cast as we know `T` is `CreateSource`
+      return (ConfiguredStatement<T>)
+          injectForCreateStatement(statement).orElse(statement);
+    } catch (final KsqlStatementException e) {
+      throw e;
+    } catch (final KsqlException e) {
+      throw new KsqlStatementException(
+          ErrorMessageUtil.buildErrorMessage(e),
+          statement.getStatementText(),
+          e.getCause());
```

Comment:
Avoid use of `ErrorMessageUtil`... it's nasty.  How about just:
```suggestion
      throw new KsqlStatementException(e.getMessage(), statement.getStatementText(), e);
```